### PR TITLE
[IOTDB-2338] ID Table recovery

### PR DESCRIPTION
**Pull request by SilverNarcissus**
## Description

Implement id table recovery logic. After this PR, id table can recover from it's log without mlog.

Design doc: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=195727793

JIRA: https://issues.apache.org/jira/browse/IOTDB-2338

<hr>

This PR has:
- [x] been self-reviewed.
    - [x] concurrent read
    - [x] concurrent write
    - [x] concurrent read and write 
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. 
- [x] added or updated version, __license__, or notice information
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTableHashmapImpl.java [hunk 1]
```diff
@@ -79,6 +79,11 @@ public IDTableHashmapImpl(File storageGroupDir) {
     }
     if (config.isEnableIDTableLogFile()) {
       IDiskSchemaManager = new AppendOnlyDiskSchemaManager(storageGroupDir);
+      try {
+        IDiskSchemaManager.recover(this);
+      } catch (IOException e) {
+        logger.error("can't recover id table of " + storageGroupDir.getName());
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTableHashmapImpl.java:85]: please print log  use {}.

**Review by wangchao316 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTableHashmapImpl.java (hunk 1 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTableHashmapImpl.java:85]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java [hunk 2]
```diff
@@ -117,14 +127,41 @@ private boolean checkLastEntry(long pos) {
...
       throw new IllegalArgumentException("can't serialize disk entry of " + schemaEntry);
     }
 
-    return 0;
+    return beforeLoc;
+  }
+
+  @Override
+  public void recover(IDTable idTable) throws IOException {
+    FileInputStream inputStream = new FileInputStream(dataFile);
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:143]: stream should try(),  or close in finally.

Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java]: fixed~

**Review by wangchao316 — APPROVED** [review]
LGTM

**Review by zyk990424 — APPROVED** [review]
LGTM~

**Review by bigreybear — APPROVED** [review]
LGTM~

**Pull request merged**